ReactAndroid/src/main/java/com/facebook/react/views/imagehelper/ResourceDrawableIdHelper.java

```diff
@@ -66,15 +86,166 @@ public int getResourceDrawableId(Context context, @Nullable String name) {
+    for (PhoneDensity phoneDensity : phoneDensities) {
+      String drawableFileParent = String.format("drawable-%s", phoneDensity.fileParentSuffix);
+      String mipMapFileParent = String.format("mipmap-%s", phoneDensity.fileParentSuffix);
+
+      String[] parentFileNames = { drawableFileParent, mipMapFileParent };
+
+      File resourceFile = checkFiles(context, parentFileNames, fileName);
+
+      if (resourceFile != null) {
+        file = resourceFile;
+      }
+
+      // If we've found a file at our current dpi level, return it.
+      // Otherwise continue looking up the chain.
+      if (densityDpi <= phoneDensity.density) {
+        if (file != null) {
+          mResourceCacheMap.put(fileName, file.getAbsolutePath());
+          return file;
+        }
+      }
+    }
+
+    // As a last resort, check the drawable/raw folders.
+    String[] parentFileNames = { "drawable", "raw" };
+    file = checkFiles(context, parentFileNames, fileName);
+
+    if (file != null) {
+      mResourceCacheMap.put(fileName, file.getAbsolutePath());
+    }
+
+    return file;
+  }
+
+  /**
+   * Given a list of files, check if any of them exist.
+   * Checks multiple extension types.
+   */
+  private File checkFiles(Context context, String[] parentFileNames, String fileName) {
+    for(String parentFileName : parentFileNames) {
+      for (String extension : RESOURCE_EXTENSIONS) {
+        File file = getFile(context, parentFileName, fileName, extension);
+        if (file.exists()) {
+          return file;
+        }
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Returns a file object with the correct directory extensions.
+   */
+  private File getFile(Context context, String parentFileName, String fileName, String extension) {
+    String fullDrawableFileName = String.format(
+      "%s/%s/%s.%s",
+      CACHE_DRAWABLE_DIRECTORY_SCHEME,
+      parentFileName,
+      fileName,
+      extension
+    );
+
+    return new File(context.getCacheDir(), fullDrawableFileName);
+  }
+
+  /**
+   * Returns the density dpi for the device.
+   */
+  private int getDensityDpi(Context context) {
+    if (densityDpi == 0) {
+      DisplayMetrics metrics = new DisplayMetrics();
+
+      WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
+      windowManager.getDefaultDisplay().getMetrics(metrics);
+
+      densityDpi = metrics.densityDpi;
+    }
+
+    return densityDpi;
+  }
+
+  enum PhoneDensity {
+    Medium(DisplayMetrics.DENSITY_MEDIUM, "mdpi"),
+    High(DisplayMetrics.DENSITY_HIGH, "hdpi"),
+    XHigh(DisplayMetrics.DENSITY_XHIGH, "xhdpi"),
+    XXHigh(DisplayMetrics.DENSITY_XXHIGH, "xxhdpi"),
+    XXXHigh(DisplayMetrics.DENSITY_XXXHIGH, "xxxhdpi");
+
+    int density;
+
+    @NonNull
+    String fileParentSuffix;
+
+    PhoneDensity(
+      int density,
+      @NonNull String fileParentSuffix
+    ) {
+      this.density = density;
+      this.fileParentSuffix = fileParentSuffix;
+    }
+  }
```

Comment:
For the sake of keeping our branch easy to rebase I would try to move *as much of this as possible* out of this file and into a stand-alone file that this drags from.

Reply:
good point, ill try to move this over to a new file.